File: frontend/webapp/src/shared/types/project.ts
import type { AssessmentCriterion } from './assessmentCriterion.ts';
import type { Version } from './version';

export const PROJECT_NAME_MIN_LENGTH = 3;
export const PROJECT_NAME_MAX_LENGTH = 64;
export const PROJECT_DESCRIPTION_MAX_LENGTH = 256;

export interface Project {
  projectId: string;
  userId: string;
  name: string;
  description: string;
  versions: Version[];
  assessmentCriteria: AssessmentCriterion[];
}

File: backend/Projects/Data/Models/Project.cs
using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using Shared.Interfaces;

namespace Projects.Data.Models;

public class Project : IUserOwned
{
    public const int MinNameLength = 3;
    public const int MaxNameLength = 64;
    public const int MaxDescriptionLength = 256;

    public required Guid ProjectId { get; set; }
    public required Guid UserId { get; set; }
    public required string Name { get; set; }
    public required string Description { get; set; }
}

internal sealed class ProjectEntityConfiguration : IEntityTypeConfiguration<Project>
{
    public void Configure(EntityTypeBuilder<Project> builder)
    {
        builder.HasKey(x => x.ProjectId);
        builder.Property(x => x.Name).HasMaxLength(Project.MaxNameLength);
        builder.Property(x => x.Description).HasMaxLength(Project.MaxDescriptionLength);
    }
}
